src/java.base/share/classes/java/text/DecimalFormat.java

```diff
@@ -2586,14 +2586,27 @@ int subparseNumber(String text, int position,
                     boolean[] stat = new boolean[STATUS_LENGTH];
                     DigitList exponentDigits = new DigitList();
 
-                    if (subparse(text, pos, "", symbols.getMinusSignText(), exponentDigits, true, stat) &&
-                            exponentDigits.fitsIntoLong(stat[STATUS_POSITIVE], true)) {
-                        position = pos.index; // Advance past the exponent
-                        exponent = (int)exponentDigits.getLong();
+                    if (subparse(text, pos, "", symbols.getMinusSignText(), exponentDigits, true, stat)) {
+                        // We parse the exponent with isExponent true, thus fitsIntoLong
+                        // only returns false here if the value exceeds Long.MAX_VALUE.
+                        // We do not need to worry about false being returned for faulty
+                        // values as they are ignored by DigitList
+                        if (exponentDigits.fitsIntoLong(stat[STATUS_POSITIVE], true)) {
+                            position = pos.index; // Advance past the exponent
+                            try {
+                                exponent = Math.toIntExact(exponentDigits.getLong());
+                            } catch (ArithmeticException ex) {
+                                // If overflow, Integer.MAX_VALUE is sufficient
+                                exponent = Integer.MAX_VALUE;
+                            }
+                        } else {
+                            // Value is greater than Long.MAX_VALUE, exponent field
+                            // is int, so assign it to the largest possible, Integer.MAX_VALUE
+                            exponent = Integer.MAX_VALUE;
+                        }
                         if (!stat[STATUS_POSITIVE]) {
                             exponent = -exponent;
                         }
-                        sawExponent = true;
```

Comment:
I see you removed this assignment. I am wondering if we need this variable at all.

Reply:
Perhaps there was a planned reason for it before, but as it stands it serves no purpose, so until there is a reason to have it, I think we can safely remove it to de-clutter the code.